src/System.Management.Automation/engine/SessionStatePublic.cs

```diff
@@ -159,6 +160,15 @@ public PSModuleInfo Module
             get { return _sessionState.Module; }
         }
 
+        /// <summary>
+        /// An <see cref="IServiceProvider"/> instance that provides custom services to other objects.
+        /// </summary>
+        public IServiceProvider ServiceProvider
+        {
+            get => _sessionState.ExecutionContext.ServiceProvider;
+            set => _sessionState.ExecutionContext.ServiceProvider = value;
+        }
```

Comment:
Having this API on `PSModuleInfo` makes it seem like any module can use this to set their own service provider. If two modules try to set up different service providers, then the last loaded will win and the other will break.

Have you considered using a `ConditionalWeakTable<PSModuleInfo, IServiceProvider>`?

Reply:
Having lower-level modules be able to rebuild a service provider with additional services may be useful.

But perhaps something that should be made more specific?

The use case I want to cover is to be able to set an initial service provider from the InitialSessionState and make it possible to have modules recreate it with some added services.

That could perhaps be abstracted with a get only IServiceProvider property, and some
```csharp
IServiceProvider ServiceProvider { get;  }
void AddServices(IServiceCollection services)
```

But in that case, I think we would have to commit to some implementation of `IServiceProvider`.

Reply:
I still think it'll be potentially challenging for different module authors to use the same instance. Especially if they find themselves in a runspace with a different implementation.

What I was getting at with the weak table suggestion, you could have a common assembly with code like this:

```csharp
namespace MyOrg
{
    public static class Services
    {
        private static readonly ConditionalWeakTable<Runspace, IServiceProvider> s_services = new();

        public static IServiceProvider Get(Runspace runspace) => s_services.GetValue(runspace, Create);

        public static IServiceProvider GetFromTLS() => s_services.GetValue(Runspace.DefaultRunspace, Create);

        private static IServiceProvider Create(Runspace key)
        {
            // etc
        }
    }
}
```

Reply:
That is quite similar to what I have been using (not with a ConditionalWeakTable, good suggestion!) 
but that doesn't really help with injecting something differently in tests, or am I missing something?

Reply:
Updated with how I'd handle a different service provider for tests. Could also just go with a static property that's like `ServiceProviderOverride`.

If that doesn't answer your question, can you provide an example of how you'd use it with the proposal for tests?

Reply:
It is possible to work around this, as you pointed out, with some static map where I could register the runspace mapped to a service provider, but then either all modules would need such a mapping, or share some common library.

By adding this to the SessionState, it seems things become much simpler.

Reply:
An alternative may be to have SessionState provide a PowerShell defined interface, where there is only a getter for the service provider, and some method like
```csharp
public interface IServices{
   IServiceProvider ServiceProvider {get;}
   void AddServicesAndRebuildProvider(IServiceCollection services);
}
```

Then there can only ever be a single ServiceProvider, but you can add services on a per-module basis.

Reply:
Yeah but I'm thinking if two modules are using a common library and they both register services of the same interface then that'll become an issue still